### PR TITLE
Front end

### PR DESCRIPTION
**Pull request by yantang-msft**
And the python web app as the frontend of atc project.
A simple data exporter is introduced, which captures incoming request of flask web framework and outgoing request of python requests library.
Also updated the docker-compose and helm files so it can be run locally or deployed to aks

### CODE REVIEW

File: frontend/data_exporter.py [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+            sanitized_parent_id += '.'
+
+        self.id = sanitized_parent_id + str(uuid.uuid4())[0:8] + "_"
+
+def before_flask_request():
+    requestId = flask.request.headers.get("Request-Id")
+    activity_id = ActivityId(requestId)
+    flask.g.activity_id = activity_id
+    flask.g.start_time = datetime.datetime.now()
+    # TODO: to make cross component work, will need to set app_id in the response header
```
Comment (karolz-ms) [inline frontend/data_exporter.py]: This comment is a bit unclear. What is app_id and where do I get it from?

File: frontend/data_exporter.py [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+    finally:
+        return
+
+def handle_flask_request_error(error):
+    # TODO: track exception and request failure
+    return
+
+# TODO: flask.g is based on request context, and doesn't work in multi-threading cases.
+# We need something like AsyncLocal, like Execution Context which is still in proposal: https://www.python.org/dev/peps/pep-0550/
+# We're currently using thread_local storage as a hack, but it's not working very well. It's like CallContext.Set() in C#, while what we need is CallContext.LogicalSet().
```
Comment (karolz-ms) [inline frontend/data_exporter.py]: What does it mean "is not working very well"? Is it "there are known bugs/limitations such as..." Or is it "we had to write more code that we wished for, but other than that there are no known issues"?

File: frontend/data_exporter.py [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+    request_id = request.headers.get("Request-Id")
+    activity_id = thread_local.dependency_activity_id_map.get(request_id)
+    start_time = thread_local.dependency_start_time_map.get(request_id)
+    end_time = datetime.datetime.now()
+    operation_name = f"{request.method} {request.path_url}"
+
+    tags = {
+        "ai.operation.id": activity_id.get_root_id(),
+        "ai.operation.parentId": activity_id.parent_id,
+        # TODO: AI SDK put this field as the flask request's operation name, but not the HTTP request
```
Comment (karolz-ms) [inline frontend/data_exporter.py]: Not sure I understand what is the remaining work here, can you rephrase?

File: k8s/atcApp/templates/airplanesvc.yaml [hunk 4]
```diff
@@ -26,7 +26,6 @@ spec:
       containers:
         - name: airplanesvc
           image: {{ .Values.container_registry }}/airplanesvc:{{ .Values.image_tag | trim }}
-          imagePullPolicy: IfNotPresent
```
Comment (karolz-ms) [inline k8s/atcApp/templates/airplanesvc.yaml]: I set the image pull policy to "Always" in my other branch--makes it easier to debug/iterate

File: k8s/deploy.sh [hunk 5]
```diff
@@ -138,12 +138,12 @@ echo "#################### Waiting for Azure to provision external IP ##########
...
+    frontendIp=$(kubectl get svc frontend -o=jsonpath="{.status.loadBalancer.ingress[0].ip}")
     if [[ $frontendIp =~ $ip_regex ]]; then
         break
     fi
     sleep 5s
 done
 
 printf "\n"
-echo "ATC service is available under http://$frontendIp:5023/api/flights"
+echo "ATC service is available under http://$frontendIp:5022"
```
Comment (karolz-ms) [inline k8s/deploy.sh]: "ATC _application_ is available..."

File: frontend/main.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+import sys
+import requests
+import json
+import threading
+import datetime
+import uuid
+import flask
+import threading
+
+# The server should be as stateless as possible. However, since monitoring flights is a long running process, we want only one process on the server at a time and notify all users when there is update.
```
Comment (karolz-ms) [inline frontend/main.py]: How about the following:

1. The server (frontend) just tracks connect/disconnect to determine if there are any clients. 
1. When first client connects, the server starts monitoring ATC service and emitting notifications via socketio
1. When the last client disconnects, the server stops monitoring ATC service and emitting notifications

Client just opens socketio connection when the user hits "Monitor" button and closes the connection when the user hits "Stop Monitoring" button

This way you do not need is_monitoring flag at all in the data model

File: frontend/data_exporter.py (hunk 1 above)
Comment (yantang-msft) [reply frontend/data_exporter.py]: Updated. It might still be hard to understand since this logic is tightly coupled to how AI work for multi components scenario.
And after looking it again, I think this is indeed an issue about configuration. Since we want to application to be unware of AI and handle that in the sidecar.

File: k8s/atcApp/templates/airplanesvc.yaml (hunk 4 above)
Comment (yantang-msft) [reply k8s/atcApp/templates/airplanesvc.yaml]: Yeah, I wasted some time during debugging because of this setting, as I always use latest as the tag instead of tagging by timestamp.
I also changed it "Always" so we won't have conflicts.

File: frontend/data_exporter.py (hunk 2 above)
Comment (yantang-msft) [reply frontend/data_exporter.py]: I would say it's currently by design in python.
We have to write more code (passing the required information when creating new thread), while for AsyncLocal and CallContext.LogicalSetData() in C#, the platform handles this for us.

File: frontend/data_exporter.py (hunk 3 above)
Comment (yantang-msft) [reply frontend/data_exporter.py]: Well, the TODO is now done. :)
I also rephrased it, hope it's clear now.

File: frontend/main.py (hunk 6 above)
Comment (yantang-msft) [reply frontend/main.py]: I like the idea that we only monitor status when it's needed.
However, the issue I'm having is the "is_monitoring" state is not persisted. And for the behavior you proposed, we also need to maintain the state "numOfConnectedClient", which will encounter the same problem.
Actually, I tried to start monitoring anyways regardless of whether there is connected client. But it's not stable even in this case, to be concrete, the client will just lost the connection and didn't get notified.
It would take me more time to figure out what's happening and make it stable.

Comment (karolz-ms) [reply frontend/main.py]: @yantang-msft So do you want me to merge this now, or are you going to investigate a bit further and try to fix the problem with the frontend losing the monitoring state? I have a bit of preference for the latter.

Other than that LGTM

Comment (yantang-msft) [reply frontend/main.py]: Let me try another base image and see if it solves the problem.

Comment (yantang-msft) [reply frontend/main.py]: I tried a pure python base images, the state is persisted. However it has other issues and are more severe... E.g., the plane somehow only fly for one stop and the status is not updated, and the logs from the python app is also not redirected.
I'd like to check this in first, since the main purpose is to show the correlation for non .net apps, but the problem here is about the python app itself when running in container.

Comment (karolz-ms) [reply frontend/main.py]: That's weird :(

let's merge and continue working on this
